cmd/containerd-shim-runhcs-v1/task_hcs.go

```diff
@@ -159,29 +165,35 @@ func newHcsTask(
 		opts.ScaleCPULimitsToSandbox = shimOpts.ScaleCpuLimitsToSandbox
 	}
 
-	system, resources, err := hcsoci.CreateContainer(ctx, &opts)
-	if err != nil {
-		return nil, err
+	if isJobContainer {
```

Comment:
It feels pretty hacky to shoehorn this logic into `hcsTask` like this. Especially given we don't use the HCS at all for a job container. Have we explored other ways of expressing this logic, such as by a new `shimPod`/`shimTask`/`shimExec` implementation?

Reply:
I had thought about it but the core logic for handling the tasks are identical, just no longer has a resources struct, we don't care about if it's a template, it can't be isolated at the moment and few other misc things. If we're fine with a bunch of copy and pasting to a new `hostProcessTask` I'm all for it

Reply:
Can we refactor the code to make the overall quality better, and remove the need to copy/paste?

Reply:
I'll play around with it now

Reply:
Ok back to this. I've tried/thought about a couple things:

1. New `hostTask` and `hostExec` types, and leaving `createPod` to handle the "what kind of pod/container are we making" logic. There actually is quite a lot you can trim off of the task implementation when you get rid of the notion that it could be hv isolated/no more LCOW to worry about, but for the most part there's a lot of c+p here. The `hostExec` implementation is worse, it's almost line for line for example. 
 
We could split out the implementation of the methods that didn't have to change at all (`forceExit, waitForExit` to name a few) to exec.go and just invoke this in both (and any future implementations of exec/task to come). 

2. Kind of a funny one, but would a rename of `hcsTask` and `hcsExec` (and I guess `wcowPodSandboxTask`) ease our weary hearts here. There's really nothing special about the lifecycle or management of host process containers that warrant new types and these containers implement all of the container/process/io interfaces we've always used for WCOW/LCOW, so the same calls can just be re-used in the same order everywhere. 

I don't like passing around `isHostProcess/jobcontainer` around very much either but there's a lot of boilerplate to make an entirely new task/exec implementation just to avoid the 20 or so lines of ugly.

Reply:
@kevpar Any thoughts here?

Reply:
> I don't like passing around isHostProcess/jobcontainer

fwiw, i found that strange when reading through here.  I don't know enough to make suggestion though.

Reply:
The more I look at this the more I think a rename of `hcsTask` and `hcsExec` might just be good in general. For instance with the work to go through a different virtstack for hyp-iso containers, we still end up down this code path also.

Reply:
(Same thing applies, there's nothing special about the task management between an hcs launched UVM and an "insert virtstack" launched one. Everything's abstracted further down.)

Reply:
> New hostTask and hostExec types, and leaving createPod to handle the "what kind of pod/container are we making" logic. There actually is quite a lot you can trim off of the task implementation when you get rid of the notion that it could be hv isolated/no more LCOW to worry about, but for the most part there's a lot of c+p here. The hostExec implementation is worse, it's almost line for line for example.

Could none of the c+p'd code be refactored for hostExec? I guess it would be easier to see any example of the implementation.

Reply:
Specifically for hcsExec from the last time I'd looked there's a couple lines you can remove from `(he *hcsExec) Kill(...)` as we don't need to care about whether it's LCOW but besides that I don't really feel like plumbing stuff out into other functions to be called in a new `jobExec` or similar is worth it.

Reply:
Another thing I thought of was, maybe we just make `jobcontainers.Create` return a resources struct also. It doesn't have the other resources the other containers types have, but it does have to do the same layer removal, with the only difference being we need to unmount the volume from the path it's at also.

Reply:
I've changed things around quite a bit here, the only place we keep a state variable to check about if it's a job container anymore is for the pod itself, due to the same IP restriction. All of the rest has gone by the wayside. I underestimated how much doing layer cleanup the same way we do it for hcs containers helped here. PTAL @kevpar